subprojects/base-services/src/main/java/org/gradle/internal/MutableReference.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.Serializable;
+
+/**
+ * A non-thread-safe type to hold a reference to a single value.
+ */
+@NotThreadSafe
+public final class MutableReference<T> implements Serializable {
+    private T value;
+
+    public static <T> MutableReference<T> empty() {
+        return of(null);
+    }
+
+    public static <T> MutableReference<T> of(@Nullable T initialValue) {
+        return new MutableReference<T>(initialValue);
+    }
+
+    private MutableReference(@Nullable T initialValue) {
```

Comment:
Why do we use a static method to instantiate `MutableReference` and constructors for `MutableBoolean`? I think we should be consistent between the two. WDYT?

Reply:
As we discussed, this is done for generics. A common pattern for the constructors is not a goal here.